lib_loader.go

```diff
@@ -0,0 +1,81 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Purego only works on linux/macOS with amd64 and arm64 from now
+//go:build (linux || darwin) && (amd64 || arm64) && !cgo
+
+package waf
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/ebitengine/purego"
+)
+
+// LibLoader is created to wraps all interactions with the purego library
+type LibLoader struct {
+	handle uintptr
+}
+
+// dlOpen open a handle for the shared library `name`.
+// the libLoader object is only a wrapped type for the linker handle
+// The `symbolsNeeded` is the list of symbols needed in the lib to load.
+// These symbols are returned via the map of symbols
+func dlOpen(name string, loader any) error {
+
+	handle, err := purego.Dlopen(name, purego.RTLD_GLOBAL|purego.RTLD_NOW)
+	if err != nil {
+		return fmt.Errorf("error opening shared library '%s'. Reason: %w", name, err)
+	}
+
+	foundHandle := false
+
+	loaderValue := reflect.ValueOf(loader).Elem()
+	loaderType := reflect.TypeOf(loader).Elem()
+	libLoader := &LibLoader{handle: handle}
+
+	for i := 0; i < loaderValue.NumField(); i++ {
+		fieldType := loaderType.Field(i)
+
+		symbolName, ok := fieldType.Tag.Lookup("dlsym")
+		if ok {
+			symbol, err := purego.Dlsym(handle, symbolName)
+			if err != nil {
+				return fmt.Errorf("cannot load symbol '%s' from library '%s'. Reason: %w", symbolName, name, err)
+			}
+
+			loaderValue.Field(i).Set(reflect.ValueOf(symbol))
+			continue
+		}
+
+		if fieldType.Type == reflect.TypeOf(libLoader).Elem() {
+			loaderValue.Field(i).Set(reflect.ValueOf(libLoader).Elem())
+			foundHandle = true
+		}
```

Comment:
So this is why you made `LibLoader` public - I was wondering why.
I suggest we keep `LibLoader` private so that we can keep the go-libddwaf public API to its bare minimum without exposing the internals.
So you can change this expected type by `io.Closer`, by embedding `io.Closer` into the `wafLoader` struct.

Reply:
If I change the embedding into `io.Closer` in `wafLoader` I do not have access to the `syscall` function in wafLoader anymore. And I don't really know if it's ok to have an interface call in a very hot path like this